satellite/satellitedb/buckets.go

```diff
@@ -0,0 +1,116 @@
+)
+
+type bucketsDB struct {
+	db dbx.Methods
+}
+
+// Buckets returns database for interacting with buckets
+func (db *DB) Buckets() buckets.DB {
+	return &bucketsDB{db: db.db}
+}
+
+// CreateBucket creates a new bucket
+func (db *bucketsDB) CreateBucket(ctx context.Context, bucket storj.Bucket) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	_, err = db.db.Create_BucketMetainfo(ctx,
+		dbx.BucketMetainfo_Id(bucket.ID[:]),
+		dbx.BucketMetainfo_ProjectId(bucket.ProjectID[:]),
+		dbx.BucketMetainfo_Name([]byte(bucket.Name)),
+		dbx.BucketMetainfo_PathCipher(int(bucket.PathCipher)),
+		dbx.BucketMetainfo_DefaultSegmentSize(int(bucket.SegmentsSize)),
+		dbx.BucketMetainfo_DefaultEncryptionCipherSuite(int(bucket.EncryptionParameters.CipherSuite)),
+		dbx.BucketMetainfo_DefaultEncryptionBlockSize(int(bucket.EncryptionParameters.BlockSize)),
+		dbx.BucketMetainfo_DefaultRedundancyAlgorithm(int(bucket.RedundancyScheme.Algorithm)),
+		dbx.BucketMetainfo_DefaultRedundancyShareSize(int(bucket.RedundancyScheme.ShareSize)),
+		dbx.BucketMetainfo_DefaultRedundancyRequiredShares(int(bucket.RedundancyScheme.RequiredShares)),
+		dbx.BucketMetainfo_DefaultRedundancyRepairShares(int(bucket.RedundancyScheme.RepairShares)),
+		dbx.BucketMetainfo_DefaultRedundancyOptimalShares(int(bucket.RedundancyScheme.OptimalShares)),
+		dbx.BucketMetainfo_DefaultRedundancyTotalShares(int(bucket.RedundancyScheme.TotalShares)),
+	)
+	if err != nil {
+		return storj.ErrBucket.Wrap(err)
+	}
+	return nil
+}
+
+// GetBucket returns a bucket from the database
+func (db *bucketsDB) GetBucket(ctx context.Context, bucketName []byte, projectID uuid.UUID) (bucket storj.Bucket, err error) {
+	defer mon.Task()(&ctx)(&err)
+	dbxBucket, err := db.db.Get_BucketMetainfo_By_ProjectId_And_Name(ctx,
+		dbx.BucketMetainfo_ProjectId(projectID[:]),
+		dbx.BucketMetainfo_Name(bucketName),
+	)
+	return convertDBXtoBucket(dbxBucket)
+}
+
+// DeleteBucket deletes a bucket
+func (db *bucketsDB) DeleteBucket(ctx context.Context, bucketName []byte, projectID uuid.UUID) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	_, err = db.db.Delete_BucketMetainfo_By_ProjectId_And_Name(ctx,
+		dbx.BucketMetainfo_ProjectId(projectID[:]),
+		dbx.BucketMetainfo_Name(bucketName),
+	)
+	return err
+}
+
+// ListBuckets returns a list of all buckets for a project
+func (db *bucketsDB) ListBuckets(ctx context.Context, projectID uuid.UUID) (buckets []storj.Bucket, err error) {
+	defer mon.Task()(&ctx)(&err)
+	rows, err := db.db.All_BucketMetainfo_By_ProjectId_OrderBy_Asc_Name(ctx,
+		dbx.BucketMetainfo_ProjectId(projectID[:]),
+	)
+	if err != nil {
+		return buckets, err
+	}
+	for _, dbxBucket := range rows {
+		item, err := convertDBXtoBucket(dbxBucket)
+		if err != nil {
+			return buckets, err
+		}
+		buckets = append(buckets, item)
+	}
+	return buckets, err
+}
+
+func convertDBXtoBucket(dbxBucket *dbx.BucketMetainfo) (bucket storj.Bucket, err error) {
+	id, err := uuid.Parse(string(dbxBucket.Id))
```

Comment:
I think this should be done more like this https://github.com/storj/storj/blob/0e528bc56ea05d896b490b3146e704492f2e9c3b/cmd/satellite/reports/attribution.go#L98